File: src/pipelines/des_y3/shear_1h2h/fast_mass/cpp/Shear1hFastMass.cc
// Miscentred one-halo shear via the fast_mass strategy — C++ backend.
//
// The exact-z-contraction fast path stated as a des_y3 module: the
// same algorithm as the production Shear1hMisSel.so (`method = exact`,
// Shear1hMisSelGL in the immutable n_operator_sel_gl_t.hh) — SelGLCore
// builds W_ij(lnM) with Sigma_crit_inv folded in, then the 1-D GL mass
// sum over the production miscentred mixture — re-expressed here with
// its own configuration label and the namespace output section
// (shear1h_fast_mass/vals, shared with the Python backend; the two are
// interchangeable and must never run in one pipeline: DataBlock
// sections do not overwrite). Every composed piece is an existing
// immutable model: SelGLCore, the lensing_weights mis helpers, the
// haloModel dSigma_nfw Interp2D and NFW_DSIGMA_MIS.
//
// Options: bin_index x r_perp cartesian grid (bin slow / R fast),
// lob_centers (default 25 37.5 52.5 130), required zt_low/zt_high/
// lnm_low/lnm_high, n_lnm (96), n_z (64).
// Status: reference re-expression; production remains Shear1hMisSel.so.
#include "models/n_operator_sel_gl_t.hh"
#include "utils/module_macros.hh"

#include <array>
#include <optional>
#include <stdexcept>
#include <vector>

class Shear1hFastMassCpp {
public:
  using grid_t = y3_cluster::grid_t<2>;
  using grid_point_t = grid_t::value_type;
  static constexpr std::size_t n_outputs = 1;

  explicit Shear1hFastMassCpp(cosmosis::DataBlock& cfg)
    : core_(cfg, module_label())
    , dsigma_mis_(4.0, 2.77533742639e+11, y3_cluster::GAMMA)
  {
    lob_centers_ =
      y3_cluster_sel_weights::mis_detail::read_lob_centers(cfg,
                                                           module_label());
    if (lob_centers_.empty())
      throw std::runtime_error("Shear1hFastMassCpp: lob_centers is empty");
  }

  void
  set_sample(cosmosis::DataBlock& s)
  {
    namespace w = y3_cluster_sel_weights;
    core_.build_weights(s, /*include_sci=*/true);
    dsigma_nfw_.emplace(y3_cluster::make_Interp2D(
      s, "haloModel", "r_sigma", "lnM", "dSigma_nfw"));
    f_mis_ = w::mis_detail::read_mis_param(s, "f_mis",
                                           w::mis_detail::F_MIS_DEFAULT);
    double const tau_mis = w::mis_detail::read_mis_param(
      s, "tau_mis", w::mis_detail::TAU_MIS_DEFAULT);
    dsigma_mis_.set_rho_mult(
      s.view<double>("cosmological_parameters", "omega_M"));
    r_mis_.assign(core_.n_bins(), 0.0);
    for (std::size_t b = 0; b != core_.n_bins(); ++b)
      r_mis_[b] = tau_mis * w::mis_detail::R_lambda(
                              lob_centers_[b % lob_centers_.size()]);
  }

  std::array<double, n_outputs>
  evaluate(grid_point_t const& pt) const
  {
    int const b = static_cast<int>(pt[0]);
    double const R = pt[1];
    if (b < 0 || static_cast<std::size_t>(b) >= core_.n_bins())
      throw std::out_of_range("Shear1hFastMassCpp: bin_index");
    auto const& wb = core_.weights(b);
    auto const& xs = core_.lnm_x();
    auto const& ws = core_.lnm_w();
    double acc = 0.0;
    for (std::size_t k = 0; k != xs.size(); ++k) {
      double const d_cen = dsigma_nfw_->clamp(R, xs[k]);
      double const d_mis = dsigma_mis_(R, r_mis_[b], xs[k]);
      acc += ws[k] * wb[k] *
             ((1.0 - f_mis_) * d_cen + f_mis_ * d_mis);
    }
    return {acc};
  }

  static char const* module_label() { return "Shear1hFastMass"; }

  static std::array<char const*, n_outputs>
  output_sections()
  {
    return {"shear1h_fast_mass"};
  }

  static grid_t
  make_grid_points(cosmosis::DataBlock& cfg)
  {
    return y3_cluster::make_grid_points_cartesian_product(
      cfg, module_label(), "bin_index", "r_perp");
  }

private:
  y3_cluster::nosel_gl_detail::SelGLCore core_;
  std::vector<double> lob_centers_;
  std::optional<y3_cluster::Interp2D> dsigma_nfw_;
  y3_cluster::NFW_DSIGMA_MIS dsigma_mis_;
  double f_mis_{y3_cluster_sel_weights::mis_detail::F_MIS_DEFAULT};
  std::vector<double> r_mis_;
};

DEFINE_COSMOSIS_SCALAR_EVALUATOR_MODULE(Shear1hFastMassCpp)

File: src/pipelines/des_y3/shear_1h2h/fast_mass/python/shear1h_fast_mass.py
"""Miscentred one-halo shear via the fast_mass strategy — Python.

The exact-redshift-contraction fast path for the shear observable,
extending the plan's `fast_mass` strategy from counts to shear at the
plan owner's direction: the z integral is done exactly, on fixed GL
nodes, OUTSIDE the radial operator —

    W_ij(lnM) = int dz  n(M,z) dV/dOmega/dz(z) Omega(z)
                        Sigma_crit_inv(z) S_ij(lnM,z)
    O_ij(R)   = int dlnM  W_ij(lnM) Phi_i(R, lnM)

with Phi_i the production miscentred mixture
(1 - f_mis) DSigma_nfw + f_mis DSigma_mis. This is exactly what the
production Shear1hMisSel.so `method = exact` computes
(Shear1hMisSelGL); here it is expressed through the shared Python
replicas — MassZWeights for W_ij (SelGLCore twin) and
lensing_profiles.MisMixtureProfile for Phi (haloModel bilinear +
NFW_DSIGMA_MIS gamma table, interpolation-exact) — so the backend
agrees with production to near machine precision (see README.md).

Unlike radial_series (whose per-(bin,R) work is three table lookups),
this pays the full n_lnm-node mass sum per (bin, R) point; unlike
full_ltmz, it consumes the tabulated S_ij rather than re-evaluating the
selection kernels.

DataBlock contract
------------------
Reads (options): bin_index, r_perp (cartesian, bin slow / R fast),
    lob_centers (default 25 37.5 52.5 130),
    zt_low, zt_high, lnm_low, lnm_high (required), n_lnm (96), n_z (64).
Reads (datablock): the full Shear1hMisSel contract —
    sel_function/{lnM,z,S_stack}, mass_function/*, cluster_abundance/*,
    distances/{z,d_a}, average_sigma_crit_inv/{zlense,sci_average},
    halomodel/{r_sigma,lnM,dSigma_nfw}, miscentering/{f_mis,tau_mis}
    (optional; defaults 0.22/0.17), cosmological_parameters/*,
    data/nfw_off_center/*gamma* (fixed tables, loaded at import).
Writes: shear1h_fast_mass/vals  (n_bins * n_r,)   [hardcoded section]

Status: reference re-expression of the production algorithm.
Production remains Shear1hMisSel.so.
"""
from __future__ import annotations

import sys
import time
from pathlib import Path

import numpy as np

for _p in Path(__file__).resolve().parents:
    if (_p / "shared" / "datablock_models.py").is_file():
        if str(_p) not in sys.path:
            sys.path.insert(0, str(_p))
        break

from shared import datablock_models as dm
from shared import lensing_profiles as lp
from systematics.selection_richness.python import sel_kernels

OUTPUT_SECTION = "shear1h_fast_mass"


def compute_shear(weights, profile, bin_index, r_perp):
    """O(R) for the requested bins: sum_k w_k W[b,k] Phi_b(R, lnM_k)."""
    n_r = len(r_perp)
    vals = np.empty(len(bin_index) * n_r)
    lnm = weights.lnm_x
    for i, b in enumerate(bin_index):
        phi = profile(b, np.asarray(r_perp)[:, None], lnm[None, :])
        vals[i * n_r:(i + 1) * n_r] = phi @ (weights.lnm_w * weights.W[b])
    return vals


def setup(options):
    from cosmosis.datablock import option_section
    sf = sel_kernels.load()
    cfg = {}
    cfg["bin_index"] = sf._read_array(options, option_section,
                                      "bin_index").astype(int)
    cfg["r_perp"] = sf._read_array(options, option_section, "r_perp")
    try:
        cfg["lob_centers"] = sf._read_array(options, option_section,
                                            "lob_centers")
    except Exception:
        cfg["lob_centers"] = np.asarray(dm.DEFAULT_LOB_CENTERS)
    for key in ("zt_low", "zt_high", "lnm_low", "lnm_high"):
        cfg[key] = float(options.get_double(option_section, key))
    for key, default in (("n_lnm", 96), ("n_z", 64)):
        try:
            cfg[key] = int(options.get_int(option_section, key))
        except Exception:
            cfg[key] = default
    return cfg


def execute(block, cfg):
    t0 = time.perf_counter()
    source = dm.DataBlockSource(block)
    weights = dm.MassZWeights(
        source, n_lnm=cfg["n_lnm"], n_z=cfg["n_z"],
        zt_lo=cfg["zt_low"], zt_hi=cfg["zt_high"],
        lnm_lo=cfg["lnm_low"], lnm_hi=cfg["lnm_high"], include_sci=True)
    profile = lp.MisMixtureProfile(
        source, lob_centers=cfg["lob_centers"],
        f_mis=dm.read_mis_param(source, "f_mis", dm.F_MIS_DEFAULT),
        tau_mis=dm.read_mis_param(source, "tau_mis", dm.TAU_MIS_DEFAULT),
        omega_m=source.scalar("cosmological_parameters", "omega_m"))

    block[OUTPUT_SECTION, "vals"] = compute_shear(
        weights, profile, cfg["bin_index"], cfg["r_perp"])
    dt_ms = 1000.0 * (time.perf_counter() - t0)
    print(f"[shear1h_fast_mass] {cfg['bin_index'].size} bins x "
          f"{cfg['r_perp'].size} radii — {dt_ms:.0f} ms", flush=True)
    return 0


def cleanup(config):
    return 0
